_posts/2018-08-30-fullstack-typed-api-with-typescript.md

```diff
@@ -0,0 +1,428 @@
+---
+layout: post
+title: Fullstack shared typed api with TypeScript
+author: jgke
+excerpt: >
+  TypeScript makes it possible to share type information between frontend and
+  backend. Here's how to use it to ensure consistent api calls between the
+  frontend and backend.
+tags:
+ - TypeScript
+ - Full stack
+---
+
+Got a yet another error due to the JavaScript frontend slipping an unnoticed
+``undefined`` to a request or passing a string instead of a number, which
+causes requests to fail without a clear reason (or worse, succeed)? Maybe you
+changed the API on the server side but forgot to update the client? Consider
+sharing the api as a single type between the frontend and the backend. This
+blog post explains the process on how to build a shared API using fullstack
+TypeScript. As a bonus, you also get runtime type checks!
+
+The source code for the finished example project is available
+[on GitHub](https://github.com/jgke/typescript-react-express-starter/).
+This post is more about the journey. For the destination, check the repository.
+
+## Sketching out an api
+
+Let's make a simple api:
+- ``GET /customers``: fetch a list of customers
+- ``POST /customers``: create a new customer
+- ``GET /search?id=123&name=foobar``: search a customer by id/name
+
+As we're building a JSON-based api, every request and return value has to be an
+object. We can handle this transparently by defining the format:
+
+```ts
+export interface ApiResponse<Res> {
+    message: Res;
+}
+```
+
+All HTTP requests look roughly like functions ``(queryParameters, body) =>
+returnvalue`` (for GETs, this means (queryParameters, {}) => returnValue``), so
+we can model the api with the following type:
+
+```ts
+type Customer = {
+    id: number,
+    name: string
+}
+type ApiMap = {
+    customers: {
+        GET: () => Promise<Customer[]>,
+        POST: (customerName: string) => Promise<Customer>
+    },
+    search: {
+        GET: (query: {id?: number, name?: string}) => Promise<Customer[]>
+    }
+}
+```
+
+On the client side, this type can be used to define a object:
+
+
+```ts
+function fetchAny(url: string, queryParams?: any,
+                  method?: string, body?: any): Promise<any> {
+    let address = new URL(url);
+    Object.keys(queryParams || {})
+        .forEach(key => address.searchParams.append(key, queryParams[key]));
+    return fetch(address.toString(), {method, body})
+        .then(response => response.json()
+            .then((responseBody: any) => new Promise<Res>(resolve =>
+                resolve(responseBody.message))));
+}
+
+const api: ApiMap = {
+    customers: {
+        GET: () => fetchAny('/customers'),
+        POST: (customerName: string) => fetchAny('/customers', {}, 'POST', customerName)
+    },
+    search: {
+        GET: (params: {id: number?, name: string?}) => fetchAny('/search', params, 'PUT')
+    }
+};
+
+// ...
+
+api.customers.GET().then(val => /* ... */ val)
+```
+
+and on the server side
```

Comment:
Oisko tästä kunnollinen lause? Jotain Now that we have the client side setup, we can implement the server side component..